### ISSUE TITLE
Vault overrides file permissions and ownership

### ISSUE THREAD

**Issue by iby** [ansible/ansible#11544]
##### ISSUE TYPE
Bug Report

##### COMPONENT NAME
vault

##### ANSIBLE VERSION
2.12

##### SUMMARY
I have several files with 400 permission. After encrypting / decrypting them they change to 644. That's not desired / expected. It seems the same happens for the file owner. If I have multiple files belonging to several users, they all become owned by the current user.

**Comment by eprigorodov** [comment]
We have been also affected by this issue. It was then partially fixed starting from [v2.0.0](https://github.com/ansible/ansible/commit/00bc74404a8d2894f303f4612353417e4e94a156).

`ansible.parsing.vault.VaultEditor` always writes results of its operation into a new temporary file, then deletes original file and renames temporary file into original name. New file inevitably belongs to a current user.

Exact code locations: [VaultEditor._shred_file()](https://github.com/ansible/ansible/blob/f8d6c71d09d33f4f54753f76801b587720412605/lib/ansible/parsing/vault/__init__.py#L659), [VaultEditor.write_data()](https://github.com/ansible/ansible/blob/f8d6c71d09d33f4f54753f76801b587720412605/lib/ansible/parsing/vault/__init__.py#L871), [VaultEditor.shuffle_files()](https://github.com/ansible/ansible/blob/f8d6c71d09d33f4f54753f76801b587720412605/lib/ansible/parsing/vault/__init__.py#L881).

There could be two possible reasons for deleting original file:

1. Security. In `ansible-vault encrypt` and `ansible-vault edit` VaultEditor takes special care to securely delete an original plaintext.
1. Atomic changes. Write error does not cause partially encrypted and/or damaged file.

The drawback is described in the issue: loss of permissions and ownership. "edit" and "rekey" operations [restore permissions](https://github.com/ansible/ansible/blob/b9762b52afac4494162935834e48cc626ff08e3c/lib/ansible/parsing/vault/__init__.py#L818) and try to [restore ownership](https://github.com/ansible/ansible/blob/b9762b52afac4494162935834e48cc626ff08e3c/lib/ansible/parsing/vault/__init__.py#L888) but the latter **only works for root**.

Security consideration is not much important: just wiping file contents without deleting the inode will give the same effect.

The second reason has working use case: running two `ansible-vault edit` operations on the same file in parallel. "encrypt", "decrypt" and "rekey" operations do not strictly implement "remove-rename" pattern and therefore do not ensure atomic change.

It seems that these precautions do not worth damage done by each operation.

I would propose to get rid of "remove-rename" pattern (i.e. from `os.remove()` calls mentioned above) and just write results into the original file.

**Comment by eprigorodov** [comment]
Correction: [rekey operation](https://github.com/ansible/ansible/blob/8d14f9d5787cd971e699ab74349d2b26cad6bbf7/lib/ansible/parsing/vault/__init__.py#L786) does not delete the original file and just writes new content over, so it is not affected by this bug.

It hints that possible reasons for deleting the original (listed above) are not very significant.

**Comment by mkrizek** [comment]
@ianbytchek @eprigorodov are you still seeing the issue with the recent versions of ansible?

needs_info

**Comment by eprigorodov** [comment]
Yes, the `devel` snapshot still does what was described above: it creates the new file, restores permissions but fails trying to restore the ownership. On Ubuntu 16, one non-root user creates the vault file and then another edits it:

    $ ansible-vault -vvv edit /opt/shared/vault.tmp
    ansible-vault 2.6.0 (devel 553cf4cdfc) last updated 2018/05/14 19:24:59 (GMT +200)
      config file = None
      configured module search path = [u'/home/<user>/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
      ansible python module location = /home/<user>/dev/ansible/lib/ansible
      executable location = /home/<user>/dev/ansible/bin/ansible-vault
      python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
    No config file found; using defaults
    Vault password: 

    ERROR! Unexpected Exception, this is probably a bug: [Errno 1] Operation not permitted: '/opt/shared/vault.tmp'
    the full traceback was:

    Traceback (most recent call last):
      File "/home/<user>/dev/ansible/bin/ansible-vault", line 118, in <module>
        exit_code = cli.run()
      File "/home/<user>/dev/ansible/lib/ansible/cli/vault.py", line 255, in run
        self.execute()
      File "/home/<user>/dev/ansible/lib/ansible/cli/__init__.py", line 155, in execute
        fn()
      File "/home/<user>/dev/ansible/lib/ansible/cli/vault.py", line 446, in execute_edit
        self.editor.edit_file(f)
      File "/home/<user>/dev/ansible/lib/ansible/parsing/vault/__init__.py", line 975, in edit_file
        force_save=False, vault_id=vault_id)
      File "/home/<user>/dev/ansible/lib/ansible/parsing/vault/__init__.py", line 884, in _edit_file_helper
        self.shuffle_files(tmp_path, filename)
      File "/home/<user>/dev/ansible/lib/ansible/parsing/vault/__init__.py", line 1096, in shuffle_files
        os.chown(dest, prev.st_uid, prev.st_gid)
    OSError: [Errno 1] Operation not permitted: '/opt/shared/vault.tmp'

The vault file then belongs to the last editor user and its default group. The edited vault file also has its permissions restored to default 0600 which makes it inaccessible for the original owner (depending on default groups).

For internal use I developed a bugfix which implements a proposal above: do not delete original vault file during encrypt/decrypt/edit operations, just overwrite old inode with the new content. I am going to send a PR tonight or tomorrow.

**Comment by ansibot** [comment]
@ianbytchek This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by eprigorodov** [comment]
@mkrizek Hi Martin, thank you for all the support in pushing this issue forward. The PR awaits a review from @alikins for a month. Is it possible to stop the bot from closing the issue automatically?

**Comment by ansibot** [comment]
@ianbytchek: Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information.

Here are the items we could not find in your description:
- ansible version

Please set the description of this issue with an appropriate template from:
https://github.com/ansible/ansible/tree/devel/.github/ISSUE_TEMPLATE

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ianbytchek You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by risadams** [comment]
This is still occurring using Anzible core version 2.14.2.

I have also noted that a (somewhat) viable workaround is to use `sudo`, which seems to leave file ownership intact.
